statistics/table.go

```diff
@@ -68,6 +70,113 @@ type Table struct {
 	// and the schema of the table does not change, we don't need to load the stats for this
 	// table again.
 	TblInfoUpdateTS uint64
+
+	IsPkIsHandle bool
+}
+
+// ColAndIndexExistenceMap is the meta map for statistics.Table.
+// It can tell whether a column/index really has its statistics. So we won't send useless kv request when we do online stats loading.
+type ColAndIndexExistenceMap struct {
+	m1            map[int64]*model.ColumnInfo
+	colAnalyzed   map[int64]bool
+	m2            map[int64]*model.IndexInfo
+	indexAnalyzed map[int64]bool
+	someAnalyzed  bool
+}
+
+// SomeAnalyzed checks whether some part of the table is analyzed.
+// The newly added column/index might not be has its stats.
+func (m *ColAndIndexExistenceMap) SomeAnalyzed() bool {
+	if m == nil {
+		return false
+	}
+	for _, v := range m.colAnalyzed {
+		if v {
+			return true
+		}
+	}
+	for _, v := range m.indexAnalyzed {
+		if v {
+			return true
+		}
+	}
+	return false
+}
+
+// Has checks whether a column/index stats exists.
+func (m *ColAndIndexExistenceMap) Has(id int64, isIndex bool) bool {
+	if isIndex {
+		_, ok := m.m2[id]
+		return ok
+	}
+	_, ok := m.m1[id]
+	return ok
+}
+
+// HasAnalyzed checks whether a column/index stats exists and it has stats.
+// TODO: the map should only keep the analyzed cols.
+func (m *ColAndIndexExistenceMap) HasAnalyzed(id int64, isIndex bool) bool {
+	if isIndex {
+		analyzed, ok := m.indexAnalyzed[id]
+		return ok && analyzed
+	}
+	analyzed, ok := m.colAnalyzed[id]
+	return ok && analyzed
+}
+
+// InsertCol inserts a column with its meta into the map.
+func (m *ColAndIndexExistenceMap) InsertCol(id int64, info *model.ColumnInfo, analyzed bool) {
```

Comment:
Do we need to add a lock to make this structure thread-safe?

Reply:
We don't need it. A simple proof is that the `map[int64]*Column` and `map[int64]*Index` also don't consider the concurrent updates.